Mathlib/Combinatorics/SimpleGraph/Metric.lean

```diff
@@ -43,105 +41,190 @@ variable {V : Type*} (G : SimpleGraph V)
 
 /-! ## Metric -/
 
+section edist
+
+/--
+The extended distance between two vertices is the length of the shortest walk between them.
+It is `⊤` if no such walk exists.
+-/
+noncomputable def edist (u v : V) : ℕ∞ :=
+  ⨅ w : G.Walk u v, w.length
+
+variable {G} {u v w : V}
+
+protected theorem Reachable.exists_walk_of_edist (hr : G.Reachable u v) :
+    ∃ p : G.Walk u v, p.length = G.edist u v :=
+  csInf_mem <| Set.range_nonempty_iff_nonempty.mpr hr
+
+protected theorem Connected.exists_walk_of_edist (hconn : G.Connected) (u v : V) :
+    ∃ p : G.Walk u v, p.length = G.edist u v :=
+  (hconn u v).exists_walk_of_edist
+
+theorem edist_le (p : G.Walk u v) :
+    G.edist u v ≤ p.length :=
+  sInf_le ⟨p, rfl⟩
+
+@[simp]
+theorem edist_eq_zero_iff_eq :
+    G.edist u v = 0 ↔ u = v := by
+  apply Iff.intro <;> simp [edist, ENat.iInf_eq_zero]
+
+theorem edist_self : edist G v v = 0 :=
+  edist_eq_zero_iff_eq.mpr rfl
+
+theorem pos_edist_of_ne (hne : u ≠ v) :
+    0 < G.edist u v :=
+  pos_iff_ne_zero.mpr <| edist_eq_zero_iff_eq.ne.mpr hne
+
+lemma edist_eq_top_of_not_reachable (h : ¬G.Reachable u v) :
+    G.edist u v = ⊤ := by
+  simp [edist, not_reachable_iff_isEmpty_walk.mp h]
+
+theorem Reachable.of_edist_ne_top (h : G.edist u v ≠ ⊤) :
+    G.Reachable u v :=
+  not_not.mp <| edist_eq_top_of_not_reachable.mt h
+
+protected theorem Connected.edist_triangle (hconn : G.Connected) :
+    G.edist u w ≤ G.edist u v + G.edist v w := by
+  obtain ⟨p, hp⟩ := hconn.exists_walk_of_edist u v
+  obtain ⟨q, hq⟩ := hconn.exists_walk_of_edist v w
+  rw [← hp, ← hq, ← Nat.cast_add, ← Walk.length_append]
+  apply edist_le
+
+theorem edist_comm : G.edist u v = G.edist v u := by
+  have {u v : V} (h : G.Reachable u v) : G.edist u v ≤ G.edist v u := by
+    obtain ⟨p, hp⟩ := h.symm.exists_walk_of_edist
+    rw [← hp, ← Walk.length_reverse]
+    apply edist_le
+  by_cases h : G.Reachable u v
+  · apply le_antisymm (this h) (this h.symm)
+  · have h' : ¬G.Reachable v u := fun h' => absurd h'.symm h
+    simp [h, h', edist_eq_top_of_not_reachable]
```

Comment:
Now that we're using ENat, we don't need to reason by cases. First, we should have this theorem in Connectivity:
```lean
theorem Walk.reverse_surjective : Function.Surjective (Walk.reverse : G.Walk u v → _) :=
  Function.RightInverse.surjective Walk.reverse_reverse
```
And then we can do some rewriting:
```suggestion
theorem edist_comm : G.edist u v = G.edist v u := by
  rw [edist_eq_sInf, ← Set.image_univ, ← Set.image_univ_of_surjective Walk.reverse_surjective,
    ← Set.image_comp, Set.image_univ, Function.comp_def]
  simp_rw [Walk.length_reverse, ← edist_eq_sInf]
```
I feel like there's a missing iInf reindexing lemma (do you know @b-mehta ?)